src/Mvc/Mvc.Razor/src/ApplicationParts/ConsolidatedAssemblyApplicationPartFactory.cs

```diff
@@ -0,0 +1,28 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Mvc.ApplicationParts
+{
+    /// <summary>
+    /// Configures an <see cref="ApplicationPart" /> that contains controllers, as well as Razor views and Pages.
+    /// <para>
+    /// Combines the results of <see cref="DefaultApplicationPartFactory.GetApplicationParts(Assembly)"/> and
+    /// <see cref="CompiledRazorAssemblyApplicationPartFactory.GetApplicationParts(Assembly)"/>. This part factory
+    /// may be used if Razor views or Razor Pages are compiled in to with other types including controllers.
+    /// </para>
+    /// </summary>
+    public sealed class ConsolidatedAssemblyApplicationPartFactory : ApplicationPartFactory
+    {
+        /// <inheritdoc />
+        public override IEnumerable<ApplicationPart> GetApplicationParts(Assembly assembly)
```

Comment:
Spicy question - why do you need this boy VS adding this functionality to `CompiledRazorAssemblyApplicationPartFactory`? I can't think of a case today where someone has controllers in their existing razor code. 

You could also simplify this other way by just building the Razor Page discovery into the default case. 

Basically what I'm saying is that on reflection, it seems like this area is overdesigned a big. Someone (wonder who's to blame) seemed to really be in love with this idea of combining codegen+metadata to configure runtime behaviour when they built this, and maybe it could just all be simpler.

Reply:
Thanks @rynowak for the feedback. It feels like either of those options has slightly higher inherent breaking change risk or work to make work compared to just adding an additional API. Particularly given that this API is a bit tucked away and MVC is a wasteland of fun APIs, this didn't seem like overkill. We'll stick with the new type for now, we have time until 6.0 ship if we start feeling differently about it.